core/src/main/scala/za/co/absa/spline/harvester/dispatcher/HttpLineageDispatcher.scala

```diff
@@ -147,9 +147,16 @@ class HttpLineageDispatcher(restClient: RestClient)
         .post(json, requestCompressionSupported)
         .throwError
         .body
-
     } catch {
-      case NonFatal(e) => throw new RuntimeException(s"Cannot send lineage data to ${endpoint.request.url}", e)
+      case e: HttpStatusException => {
+        log.error(s"Could not send lineage data to ${endpoint.request.url}: ${e.code} ${e.statusLine}", e)
+        log.debug(s"Error response body: ${e.body}")
+        throw new RuntimeException(s"Cannot send lineage data to ${endpoint.request.url}", e)
```

Comment:
Logging and throwing at the same time is considered an anti-pattern.
Throwing means giving up and delegating the error handling responsibility to the caller. In this case you don't know if it's actually an error at all, because you don't know how the calling is going to handle it. 

If the intention is to log more details on exception then the better approach would be either adding that info into the exception message, or (if you want to be picky and type safe) creating another exception class with corresponding fields.

Reply:
that's great example when trying to avoid "antipattern" caused several days of debugging instead of 5 minutes just because this log wasn't here :)

Reply:
i.e. that exception was handled by spark by just logging it as warning with message completely not related to spline - it was very easy to miss it because it was not logged by spline at all

I added that log here completely intentionally because of this reason - i still think that it will make life easier for future users if they have to debug some problems

Reply:
Sorry I cannot agree with this statement. It's wasn't avoiding anti-patterns that caused you several days of debugging, but missing logs. I'm not again adding logs, don't get me wrong. My point is to do it in a right way.

Reply:
I've added logs to spline logger which is expected to log problems about http communication `HttpLineageDispatcher`

Adding message details to exception itself would not help at all because it was caught by spark internals and logged as spark internals logs thus making it more difficult to filter all logs from spline - that's exactly what caused those problems, spline did not log communication error details at all, extending exception message won't help at all, catching HTTP exceptions in different place also not the best idea, not throwing exception even worse - thus i still think that's the best way to have this situation improved - we have exact HTTP problem logs in http related component and still throwing exception as unexpected situation

I have no plans to make changes to this PR, so if you don't agree we can just close it, but it will be still difficult to debug http communication from agent

As I can see i'm not the only one with problems https://github.com/AbsaOSS/spline/issues/683#issuecomment-634762779

Reply:
thank you @DaimonPl for you contribution. I'll give it another look.